[6.1] Batch Copy & Move tags

beni71 opened the pull request:
Pull Request for Issue #23304, only partially.
This PR is based on #38583 and fixes the comments.

### Summary of Changes
Adds the ability to batch copy/move tags in the tags list.
Copying a tag also copies its children (same behavior as in other tree models like categories, menues).
Moving a tag also moves its children (same behavior as in other tree models like categories, menues).
For example, it is possible to move a child tag from one parent to another (including its children), or to copy a child tag from one parent to another (including its children).
In the batch tag drop down there is also an item "Root" to move/copy selected tags to the root level.

Changed visualization "Action to Perform" of copy/move action, it is now hidden as long as the user does not touch the select field for copy/move. See example in video below. This change affects batch copy/move dialog of menues and articles as well.

It also fixes a browser console error when the batch dialog is opened and cancel is clicked, on systems where multilingual is disabled.

### Testing Instructions
It contains an updated javascript so run: npm run build:js

Initial tag structure **for each** of the tests below:
Tag1
  -Tag2
  --Tag3
Tag4

**Test 1**
Move Tag2 to Tag4 results in:
Tag1
Tag4
  -Tag2
  --Tag3

**Test 2**
Move Tag1 to Tag4 results in:
Tag4
  -Tag1
  --Tag2
  ---Tag3

**Test 3**
Copy Tag4 to Tag2 result in:
Tag1
  -Tag2
  --Tag3
  --Tag4 (2)
Tag4

**Test 4**
Check whether the visualization of "Action to Perform" of copy/move action is first hidden for views like menues, articles. Only if a target item is selected "Action to Perform" appears.

### Actual result BEFORE applying this Pull Request
The tags list batch processing does not support copy move.


### Expected result AFTER applying this Pull Request
The tags list batch processing supports copy move:

![tag-copy-move](https://github.com/joomla/joomla-cms/assets/8353850/01aa02bd-44f2-4728-b5d8-5988be271011)


### Link to documentations
Please select:
- [ ] Documentation link for docs.joomla.org: <link>
- [ ] No documentation changes for docs.joomla.org needed

- [ ] Pull Request link for manual.joomla.org: <link>
- [ ] No documentation changes for manual.joomla.org needed

ceford commented:
I have tested this item :white_check_mark: successfully on c503b035749dca4acc30300fc3699058d75f22c3<br /><br />Easy to get into a muddle but works as described.<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/41927">issues.joomla.org/tracker/joomla-cms/41927</a>.</sub>

brianteeman commented:
@chmst isnt this **identical** to the ui for batch copying and moving menu items??

chmst commented:
@brianteeman not if I understand this correctly

You choose a menuitem in the list (the systems hilds the id), opens the batch module 

![grafik](https://github.com/joomla/joomla-cms/assets/1035262/5ff5d6d6-fe68-49b6-b074-6a0417ca33da)

then you choose a target menuitem and decides if you want a copy or move.

You must not do this first step and select our item in the batch modal item in the batch modal.

brianteeman commented:
![image](https://github.com/joomla/joomla-cms/assets/1296369/2833d4ef-86e6-4190-a24f-9cce1aa9a5b2)

![image](https://github.com/joomla/joomla-cms/assets/1296369/7e22aeff-61ef-4812-9c9d-c2946cd71277)

![image](https://github.com/joomla/joomla-cms/assets/1296369/4d2426dc-a871-4e4b-8e22-661ce5922b31)

![image](https://github.com/joomla/joomla-cms/assets/1296369/cec9fdc2-3f1f-4a55-aab9-6408360a1f80)

I dont see the difference

I think the confusion is because the language string is wrong
it should **not** be _**Keep Original Tags**_
it should be _**Dont Copy or move**_

![image](https://github.com/joomla/joomla-cms/assets/1296369/c947e013-1b1e-4d0d-87ba-b05ddbf8d2ec)

chmst commented:
Yes, thanks for explainingt. The language string is the first reason for confusion. 

The second is: Why not accept the selection of items to move or copy from the overview? In com_menus and in categories we don't have an extra step for that. Maybe there is something special in tags which requires that?

brianteeman commented:
@chmst Sorry I dont understand what you are talking about

chmst commented:
I am sorry that I cannot explain. Will try it later with videos.

chmst commented:
@brianteeman you were correct. - it seems that this language string caused the whole confusion and my folloing action was overreaction. 
So it will be an easy fix. Thanks for your patience.

beni71 commented:
Sorry for that confusion. I fixed the text and updated the description part of the PR.

HLeithner commented:
This pull request has been automatically rebased to 5.1-dev.

Quy commented:
Please fix phpcs https://ci.joomla.org/joomla/joomla-cms/71117/1/7

steffans commented:
I have tested this item :red_circle: unsuccessfully on 78aad67cbc53c14c610e9a7627ddfdac2ade5bf0<br /><br />My result for Test 2 looks like this, which is what i would expect. But its not mentioned in the test instructions.

**Test 2** 
Tag4
-Tag2
--Tag3
-Tag1

Fails to copy the tags recursively, somehow the Tag 4 is added 2x, see Tag4 (15) and --Tag4 (8)?

**Test 3**
Tag4
-Tag2
--Tag3
--Tag4 (8)
---Tag2 (8)
----Tag3 (5)
----Tag4 (15)
---Tag1 (2)
-Tag1
<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/41927">issues.joomla.org/tracker/joomla-cms/41927</a>.</sub>

exlemor commented:
I have tested this item :red_circle: unsuccessfully on 78aad67cbc53c14c610e9a7627ddfdac2ade5bf0<br /><br />When I try to do the first move of -Tag2 --Tag3 under Tab4, I do not even have a Move or Copy dialog/set of buttons underneath the "To Move or Copy your selection please select a Tag or parent item" dropdown.

:(<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/41927">issues.joomla.org/tracker/joomla-cms/41927</a>.</sub>

steffans commented:
@exlemor You need to compile the Javascript using "npm ci" in order to test this PR

crimle commented:
I have tested this item :white_check_mark: successfully on 78aad67cbc53c14c610e9a7627ddfdac2ade5bf0<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/41927">issues.joomla.org/tracker/joomla-cms/41927</a>.</sub>

adj9 commented:
The process works, but in the batch a way to override the parent tag would be useful. Take it to the rool level<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/41927">issues.joomla.org/tracker/joomla-cms/41927</a>.</sub>

Hackwar commented:
@beni71 Thank you for your work, however it looks as if we need some more adjustments. It seems as if the hierarchy is not maintained? Could you have a look?

beni71 commented:
Hi all. Thanks for testing this feature. There are a couple of things I have to go one by one.

@steffans I'm not sure whether you didn't exactly follow the test instructions or you really found an issue. So let's go into details of test2 first:
The initial tag structure is:
Tag1
-Tag2
--Tag3
Tag4

Which means that Tag1 an Tag4 are both on root level, Tag2 is a sub-tag of Tag1 and Tag3 is a sub-tag of Tag2.
So when I move Tag1 to Tag4 then Tag1 should be a sub-tag of Tag4, like:
Tag4
-Tag1
--Tag2
---Tag3

Can you tell me how your repro steps were to get:
Tag4
-Tag2
--Tag3
-Tag1

beni71 commented:
Extended the batch tag drop down with a "Root" item to move/copy selected tags to the root level, suggested by @adj9.
I updated the description.

beni71 commented:
@bembelimen Can you please tell me where the breaking change is? Do I need to fix it?

HLeithner commented:
This pull request has been automatically rebased to 5.2-dev.

cybersalt commented:
I have tested this item :red_circle: unsuccessfully on 715e848376c32ac1d3c1678762f1c9afa17234e3<br /><br />The copy and move radio buttons did not show up for me.<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/41927">issues.joomla.org/tracker/joomla-cms/41927</a>.</sub>

alikon commented:
did you run: `npm run build:js` ?

fgsw commented:
Download Prebuilt packages result in a 404.

fgsw commented:
I have tested this item :white_check_mark: successfully on 715e848376c32ac1d3c1678762f1c9afa17234e3<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/41927">issues.joomla.org/tracker/joomla-cms/41927</a>.</sub>

bascherz commented:
I have tested this item :red_circle: unsuccessfully on 715e848376c32ac1d3c1678762f1c9afa17234e3<br /><br />The Copy/Move radio buttons do not appear. The action performed is Move.<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/41927">issues.joomla.org/tracker/joomla-cms/41927</a>.</sub>

cybersalt commented:
> I have tested this item 🔴 unsuccessfully on [715e848](https://github.com/joomla/joomla-cms/commit/715e848376c32ac1d3c1678762f1c9afa17234e3)The copy and move radio buttons did not show up for me.
> 
> This comment was created with the [J!Tracker Application](https://github.com/joomla/jissues) at [issues.joomla.org/tracker/joomla-cms/41927](https://issues.joomla.org/tracker/joomla-cms/41927).

I just learned what an NPM patch is and that it can't be tested through the patchtester as I was using it on a copy of my live site on the server. So please disregard my failed test!

obuisard commented:
> I have tested this item 🔴 unsuccessfully on [715e848](https://github.com/joomla/joomla-cms/commit/715e848376c32ac1d3c1678762f1c9afa17234e3)The Copy/Move radio buttons do not appear. The action performed is Move.

Did you use the patch tester extension for your tests? Pull requests that are tagged 'NPM Resource Changed' need to be tested locally, they cannot be tested straight from our PBF instances, unfortunately. That's because a script file has been modified and the Joomla instance needs to be rebuilt to be able to get tested.

fgsw commented:
> I just learned what an NPM patch is and that it can't be tested through the patchtester

@cybersalt you can use instead a prebuilt package. Download at the end of this pull request.

> So please disregard my failed test!

if you don't test again you can set back the test result at the issue tracker by "Test this > Not tested".

cybersalt commented:
I have not tested this item.<br /><br />My Pizza Bugs and Fun test was first marked as a fail, but then I learned that I could not test this one with the patch tester - so I am resetting this to "not tested" so as to not confuse things.<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/41927">issues.joomla.org/tracker/joomla-cms/41927</a>.</sub>

HLeithner commented:
This pull request has been automatically rebased to 5.3-dev.

web54 commented:
I have tested this item :white_check_mark: successfully on 715e848376c32ac1d3c1678762f1c9afa17234e3<br /><br />I tested it succesfully during PBF 2025<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/41927">issues.joomla.org/tracker/joomla-cms/41927</a>.</sub>

HLeithner commented:
This pull request has been automatically rebased to 6.0-dev.

QuyTon commented on administrator/components/com_tags/src/Model/TagModel.php:
```diff
@@ -403,11 +430,287 @@ protected function generateNewTitle($parentId, $alias, $title)
+     * @param   array    $pks       An array of row IDs.
+     * @param   array    $contexts  An array of item contexts.
+     *
+     * @return  mixed  An array of new IDs on success, boolean false on failure.
```
Please update. See #42972 #43006.

richard67 commented:
@Hackwar Maybe you can advice with the PHPstan errors for this PR: One can be fixed with `$this->getDocument()`, all the others are all related to the deprecated `getError()` method. What should be done with them? Change this PR's error handling to catch exceptions? Or add them to the PHPstan baseline with this PR? If we get the PR fixed we can add back the "PBF" label.

Denitz committed:
[5.4] Update NULLs in Smart Search links (#46395)

Bug fix: Articles that originally expired cannot be found on smart search even the  finish publishing date is deleted

travisrisner committed:
[5.4] Generate unique alias on “Save as Copy” for Smart Search filters (#46081)

joomla-translation-bot committed:
[5.4] Translation Update (#46407)

* Only the file `cs-CZ/joomla.cli.ini` has changed

tkuschel committed:
[5.4] Fix xml sha verification (#44336)

pe7er committed:
[5.4] Fix for untranslated JLIB_APPLICATION_ERROR string (#45028)

AkameOuO committed:
[5.4] Fix headers are not sent correctly in com_media api controller (#45150)

alikon committed:
[5.4] Task SessionGC delete metadata from db (#46128)

Fedik committed:
[5.4] Error page fix: prevent modules rendering in the incomplete Application (#46272)

bembelimen committed:
[5.4] Fix issues in language installer (#46403)

* Hide autoupdate box when installing languages

* Fix loader when installing the language

* Fix auto update disabled

brianteeman committed:
[5.4] fix version (#46425)

Due to a typo the version was not set correctly in #45762 and #45801

Signed-off-by: BrianTeeman <brian@teeman.net>

nordmograph committed:
[5.4] Do not get the profile info for the linked user when no user is linked (#46398)

nordmograph committed:
[5.4] Update repeatable-table.php (#46287)

* Update repeatable-table.php

The <th> width was hardcoded to 45%, regardless of the number of columns.
This change makes it dynamic, distributing the width according to the number of fields, while reserving 8% for the action column.

brianteeman committed:
[5.4]codestyle comment block (#46427)

Just for consistency - code review only

Signed-off-by: BrianTeeman <brian@teeman.net>
Co-authored-by: Richard Fath <richard67@users.noreply.github.com>

richard67 committed:
[5.4] NPM audit fix indirect dev dependencies (#46429)

Hackwar committed:
[5.4] Preventing notices for broken images (#46435)

richard67 committed:
Joomla! 5.4.1 Release Candidate 1

richard67 committed:
Reset to dev

beni71 commented:
@QuyTon I'm sorry, but could it be that I have reverted your last commit (rebase to 6.1)? If so, this was not my idea. Can you please take a look?

richard67 commented:
> @QuyTon I'm sorry, but could it be that I have reverted your last commit (rebase to 6.1)? If so, this was not my idea. Can you please take a look?

@beni71 I'm not Quy, but I will try to help. The problem seems to be caused a force-push of your branch.

It could be a bit complicated to fix that.

I can try, but maybe it would be easier for you to redo the PR again, create a new branch based on the 6.1-dev branch of the CMS repo (upstream) and then apply your changes again?

Let me know if you can do that or if I shall try to fix this PR here.
